File: code/5._Huffman_Coding/PHP/programming.php/Huffman_Coding.php
<?php
class HuffmanNode {
    public $data;
    public $frequency;
    public $left;
    public $right;

    function __construct($data, $frequency) {
        $this->data = $data;
        $this->frequency = $frequency;
        $this->left = null;
        $this->right = null;
    }
}

function buildHuffmanTree($freqMap) {
    $minHeap = new SplMinHeap();

    foreach ($freqMap as $char => $freq) {
        $node = new HuffmanNode($char, $freq);
        $minHeap->insert($node);
    }

    while ($minHeap->count() > 1) {
        $left = $minHeap->extract();
        $right = $minHeap->extract();

        $merged = new HuffmanNode(null, $left->frequency + $right->frequency);
        $merged->left = $left;
        $merged->right = $right;

        $minHeap->insert($merged);
    }

    return $minHeap->top();
}

function generateHuffmanCodes($root, $code, &$huffmanCodes) {
    if ($root === null) {
        return;
    }

    if ($root->data !== null) {
        $huffmanCodes[$root->data] = $code;
    }

    generateHuffmanCodes($root->left, $code . "0", $huffmanCodes);
    generateHuffmanCodes($root->right, $code . "1", $huffmanCodes);
}

function huffmanCoding($text) {
    $freqMap = [];

    for ($i = 0; $i < strlen($text); $i++) {
        $char = $text[$i];
        if (array_key_exists($char, $freqMap)) {
            $freqMap[$char]++;
        } else {
            $freqMap[$char] = 1;
        }
    }

    $root = buildHuffmanTree($freqMap);
    $huffmanCodes = [];
    generateHuffmanCodes($root, "", $huffmanCodes);

    $encodedText = "";
    for ($i = 0; $i < strlen($text); $i++) {
        $encodedText .= $huffmanCodes[$text[$i]];
    }

    return $encodedText;
}

$text = "this is an example for huffman encoding";
$encodedText = huffmanCoding($text);

echo "Huffman Codes:\n";
foreach ($huffmanCodes as $char => $code) {
    echo "$char: $code\n";
}
echo "Encoded Text: $encodedText\n";
?>

File: code/5._Huffman_Coding/C#/programming.cs/Huffman_Coding.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

class HuffmanNode : IComparable<HuffmanNode>
{
    public char Data { get; set; }
    public int Frequency { get; set; }
    public HuffmanNode Left { get; set; }
    public HuffmanNode Right { get; set; }

    public int CompareTo(HuffmanNode other)
    {
        return Frequency - other.Frequency;
    }
}

class HuffmanCoding
{
    public static HuffmanNode BuildHuffmanTree(Dictionary<char, int> freqMap)
    {
        var minHeap = new PriorityQueue<HuffmanNode>();

        foreach (var entry in freqMap)
        {
            minHeap.Enqueue(new HuffmanNode { Data = entry.Key, Frequency = entry.Value });
        }

        while (minHeap.Count > 1)
        {
            var left = minHeap.Dequeue();
            var right = minHeap.Dequeue();

            var merged = new HuffmanNode { Frequency = left.Frequency + right.Frequency, Left = left, Right = right };

            minHeap.Enqueue(merged);
        }

        return minHeap.Dequeue();
    }

    public static void GenerateHuffmanCodes(HuffmanNode root, string code, Dictionary<char, string> huffmanCodes)
    {
        if (root == null)
            return;
        if (root.Left == null && root.Right == null)
        {
            huffmanCodes[root.Data] = code;
        }
        GenerateHuffmanCodes(root.Left, code + "0", huffmanCodes);
        GenerateHuffmanCodes(root.Right, code + "1", huffmanCodes);
    }

    public static string HuffmanCoding(string text)
    {
        var freqMap = new Dictionary<char, int>();
        foreach (var c in text)
        {
            if (freqMap.ContainsKey(c))
                freqMap[c]++;
            else
                freqMap[c] = 1;
        }

        var root = BuildHuffmanTree(freqMap);
        var huffmanCodes = new Dictionary<char, string>();
        GenerateHuffmanCodes(root, "", huffmanCodes);

        var encodedText = new StringBuilder();
        foreach (var c in text)
        {
            encodedText.Append(huffmanCodes[c]);
        }

        return encodedText.ToString();
    }

    public static void Main()
    {
        var text = "this is an example for huffman encoding";
        var encodedText = HuffmanCoding(text);
        Console.WriteLine("Huffman Codes:");
        foreach (var entry in huffmanCodes)
        {
            Console.WriteLine(entry.Key + ": " + entry.Value);
        }
        Console.WriteLine("Encoded Text: " + encodedText);
    }
}
